instrumentation/rxjava-2.0/library/src/main/java/io/opentelemetry/instrumentation/rxjava2/TracingAssembly.java

```diff
@@ -0,0 +1,338 @@
+            }));
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private static void enableSingle() {
+    oldOnSingleSubscribe = RxJavaPlugins.getOnSingleSubscribe();
+    RxJavaPlugins.setOnSingleSubscribe(
+        biCompose(
+            oldOnSingleSubscribe,
+            new BiConditionalOnCurrentContextFunction<Single, SingleObserver>() {
+
+              @Override
+              SingleObserver applyActual(
+                  Single single, SingleObserver singleObserver, Context ctx) {
+                try (Scope ignored = ctx.makeCurrent()) {
+                  return new TracingSingleObserver(singleObserver, ctx);
+                }
+              }
+            }));
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private static void enableMaybe() {
+    oldOnMaybeSubscribe = RxJavaPlugins.getOnMaybeSubscribe();
+    RxJavaPlugins.setOnMaybeSubscribe(
+        (BiFunction<? super Maybe, MaybeObserver, ? extends MaybeObserver>)
+            biCompose(
+                oldOnMaybeSubscribe,
+                new BiConditionalOnCurrentContextFunction<Maybe, MaybeObserver>() {
+
+                  @Override
+                  MaybeObserver applyActual(Maybe maybe, MaybeObserver maybeObserver, Context ctx) {
+                    try (Scope ignored = ctx.makeCurrent()) {
+                      return new TracingMaybeObserver(maybeObserver, ctx);
+                    }
+                  }
+                }));
+  }
+
+  private static void disableConnectableFlowable() {
+    RxJavaPlugins.setOnConnectableFlowableAssembly(oldOnConnectableFlowableAssembly);
+    oldOnConnectableFlowableAssembly = null;
+  }
+
+  private static void disableConnectableObservable() {
+    RxJavaPlugins.setOnConnectableObservableAssembly(oldOnConnectableObservableAssembly);
+    oldOnConnectableObservableAssembly = null;
+  }
+
+  private static void disableParallel() {
+    RxJavaPlugins.setOnParallelAssembly(oldOnParallelAssembly);
+    oldOnParallelAssembly = null;
+  }
+
+  private static void disableObservable() {
+    RxJavaPlugins.setOnObservableSubscribe(oldOnObservableSubscribe);
+    oldOnObservableSubscribe = null;
+  }
+
+  private static void disableCompletable() {
+    RxJavaPlugins.setOnCompletableSubscribe(oldOnCompletableSubscribe);
+    oldOnCompletableSubscribe = null;
+  }
+
+  private static void disableFlowable() {
+    RxJavaPlugins.setOnFlowableSubscribe(oldOnFlowableSubscribe);
+    oldOnFlowableSubscribe = null;
+  }
+
+  private static void disableSingle() {
+    RxJavaPlugins.setOnSingleSubscribe(oldOnSingleSubscribe);
+    oldOnSingleSubscribe = null;
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private static void disableMaybe() {
+    RxJavaPlugins.setOnMaybeSubscribe(
+        (BiFunction<? super Maybe, MaybeObserver, ? extends MaybeObserver>) oldOnMaybeSubscribe);
+    oldOnMaybeSubscribe = null;
+  }
+
+  private abstract static class ConditionalOnCurrentContextFunction<T> implements Function<T, T> {
```

Comment:
Ah I have a feeling you're following the pattern in the Armeria instrumentation - then yeah my if-statement is similar to their `mapCurrent`. But I guess we don't really need it since context storage short circuits for `root`

Reply:
Sure, I will get rid of the helper classes.